### ISSUE TITLE
Multilevel template inheritance not working properly

### ISSUE THREAD

**Issue by schirliuradu** [maizzle/framework#98]
- Maizzle Version: 0.7.3
- Node.js Version: 10.18.1

### Description:
Hi. I want to thank you for first for having done for all of us maizzle framework. It's an awesome tool for writing email templates. 

However, there is a thing i would like to report to you. I saw that this argument was already discussed in an other issue some months ago [#69](https://github.com/maizzle/framework/issues/69), and you've found already a solution, but this seems not to work properly. 

The problem occurs when i try to extend three or more templates. Take this example: 
i have the default layout, that is extended by a first template (first.njk)
```
---
layout: 'src/layouts/default.njk'
level: 1
---

{% block template %}
  <table class="w-600 sm:w-full">
    <tr>
      <td class="p-48 sm:py-32 sm:px-24 text-center">
        {% block level %}
          {{ page.level }}
        {% endblock %}

        {% block content %}
            first level content
          {% endblock %}
      </td>
    </tr>
  </table>
{% endblock %}
```
Then i have a second template, called "second.njk", that extends the first
```
---
layout: 'src/templates/first.njk'
level: 2
---

{% block content %}

  second level content that overwrites the  first one {{ page.level }}

{% endblock %}
```
Here everything is still working, but when you try to add a third template (third.njk), that extends the second, there are some obvious problems
```
---
layout: 'src/templates/second.njk'
level: 3
---

{% block content %}

  third level content that overwrites the second one {{ page.level }}

  {% block other_block %}
    new
  {% endblock %}

{% endblock %}
```

The first problem is that i want to extend the template and define everytime just the "content" block. I don't want to define everytime the "template" block, because, in this case, i have there just some layout (markup and css) and i don't want to rewrite it everytime. Rewriting the "content" block in "second.njk" works, and also the layout (markup and css) from "first.njk" is there. But when i overwrite the "content" block in "third.njk", the layout is no more there. 

The second problem is that i can't use `{{ super() }}` in the blocks: let's say i want to take everything from "content" block of "first.njk" also in "second.njk", define something else there and then take all this content in "third.njk" with another `{{ super }}`, defining even here something else but keeping everything was already done in other templates. This doesn't work. 

The reason of why this is not working is in "toString.js": 

```
    html = layout ? `${blockStart} extends "${layout}" ${blockEnd}\n${html}` : html
    html = nunjucks.renderString(html, { page: config, env: options.env, css: compiledCSS })

    while (fm(html).attributes.layout) {
      const front = fm(html)
      html = `${blockStart} extends "${front.attributes.layout}" ${blockEnd}\n${blockStart} block template ${blockEnd}${front.body}${blockStart} endblock ${blockEnd}`
      html = nunjucks.renderString(html, { page: config, env: options.env, css: compiledCSS })
    }
```

Maizzle here is doing something that nunjucks already comes with, in another way, that causes the problems that i mentioned above. 

The solution would be to remove the while loop, and to use the `{% extends "template" %}` syntax, right after the definition of the front matter. This works, inheritance works very well, and i can call `{{ super() }}` in every inheritance level, in the same block. I could also define just the blocks that i want, in every level, without having to define the "template" block everytime. Just this would be enough to solve

```
    html = layout ? `${blockStart} extends "${layout}" ${blockEnd}\n${html}` : html
    html = nunjucks.renderString(html, { page: config, env: options.env, css: compiledCSS })
```

In this case, who is using just one level of inheritance (template that extends the layout), could omit the `{% extends "template" %}` syntax, and who's doing multiple levels of inheritance should use it for all the levels. 

In my opinion nunjucks is a very, very powerful tool, and maizzle should use all the best features that nunjucks offers. The multilevel inheritance gives me the possibility to set the "skeleton" of my email template just once, and change then the content of the emails that i have to send.

**Comment by cossssmin** [comment]
Hi,

The while loop is there so that multilevel inheritance can reach the default layout extended. Without it, you would get an empty body, we tried that. Please see [this comment](https://github.com/maizzle/framework/issues/69#issuecomment-565003979) and then the convo [from here on](https://github.com/maizzle/framework/issues/69#issuecomment-565035360). 

The trade-off is that you need to use `{% block template %}` so that the default layout feature is properly used with multilevel inheritance.

**Comment by cossssmin** [comment]
Reality is gotchas like these are hard to spot without a proper test suite, which is the next thing I need to focus on so we can get the first major release out the door 👍

**Comment by schirliuradu** [comment]
Yes, i know, i've already read all that issue, but the problem persists. I think you should just give a try to my example, and you will see that there is something that does not work fine. Otherwise, i would just to see and example splitted on 4 levels that works fine.

We solve the problem if we remove the while loop and use the `{% extends %}` syntax. If we remove the while loop and keep specifiing the layout in the front matter without using the `{% extends %}` syntax doesn't work, i agree with that. 

It seems also to work without removing the while loop and using the `{% extends %}` syntax after the front matter, but there should be also the layout in the front matter: without it is doesn't work.

**Comment by cossssmin** [comment]
Does the problem exist when using `{% block template %}` as [documented](https://maizzle.com/docs/templates/#inheritance-gotchas)?

**Comment by j-mori** [comment]
Hi guys, i have the same problem, after 3rd childe, inheritance broke and any reference to 1st template is broken. using `{% block template %} ` doesn't solve my issue.

I found a (pretty bad) workaround using both, frontMatter layouts and nunjack extends, it will work:

```
---
layout: src/templates/test2.njk
level: 3
---
{% extends 'src/templates/test2.njk'  %}
{% block test3 %}test{{page.level}}{% endblock %}
```

But as you can guess this is not an optimal solution.

**Comment by cossssmin** [comment]
@j-mori same question/suggestion as above - does it work with `{% block template %}`? You have `{% block test3 %}` and it can't work, this is already documented and the reasoning explained in #69  (trade-off so that we can still use a 'default' layout, which is one of Maizzle's features).

**Comment by schirliuradu** [comment]
I read that part of the documentation. Even if you use `{% block template %}` in all the templates, you can not use super more than on one level inside that block, you should try it. You lose also the parts of that block that you defined in previous templates in that block. For keeping those parts you should use super, that doesn't work, or you should define again everything, which has no sense.  

I think the best solution would be to let all the compiling stuff to nunjucks, that works great and has nice features. Using it's inheritance maizzle becomes super powerful.

**Comment by cossssmin** [comment]
OK, I'll have another look 👍

How do you suggest we keep the functionality for the [default layout](https://maizzle.com/docs/layouts/#default-layout) though?

**Comment by cossssmin** [comment]
I created a test repo [here](https://github.com/maizzle/test-inheritance), clone it and try it out. I've added `{{ super() }}` in the third layout, you can see the compiled templates in `build_production`.

If you comment out the `while` loop in `node_modules/@maizzle/framework/src/generators/output/toString.js` and then run `maizzle build production`, you'll see the third template not using the `src/layouts/default.njk` layout anymore, and front matter is not being parsed & removed.

How do you suggest we handle this?

**Comment by cossssmin** [comment]
I can see how the content from the first template is lost in the third if we keep the while loop, and I'd love to get this fixed. 

Maybe I'm missing something from your example, could you please put up a repo that shows how you have it working, so I can reproduce by removing the loop?

**Comment by schirliuradu** [comment]
Hi Cosmin, 

Yes, of course. You have two examples [here](https://github.com/schirliuradu/maizzle-examples):

1. maizzle-original-compiler - in this one, i kept maizzle actual default behaviour. This is the same example that you have in your repo. You added `{% super() %}` in "three.njk", and it seems to work, at least there are no errors. In my example, i added just `{% super() %}` also in "two.njk", and this causes an error: it doesn't work. 

2. maizzle-nunjucks-compiler - here i have a real example, in which you can see all the power of inheritance, and it is also very useful for many real life cases, in my opinion. In "toString.js" i commented the parts without which everything is working. Test this example, please, and you will see that it would be very nice to have and use all the potential that nunjucks offers us. In this second example, i use as many  `{% super() %}` as i want, and this is fine. There is just a problem, of the default layout: you have to specify the layout that you extend with the `{% extends %}` syntax. You can still have here the front matter, if you want to set other variables of course. 

I don't have a solution for the default layout at the moment, but i want just to show you the potential of nunjucks inheritance in my second example, which used in maizzle would be amazing. 

Try it, and let me know.

**Comment by cossssmin** [comment]
Thanks Radu, I'll have a look later today when I get some time.

What I find odd is that it would work with the `{% extends %}` in the template, but not if we define basically exactly the same thing in `toString()`. I'll need to have a closer look at that `layout` variable there, because it should theoretically work either way...

**Comment by cossssmin** [comment]
Thanks for the example Radu, I see it working properly indeed in your example. No idea why this:

https://github.com/maizzle/framework/blob/9ee6bac7f96e831c6b30eaab82e067a83562164d/src/generators/output/toString.js#L61

doesn't work though, my assumption is Nunjucks loads the extended layout/template before we can tell it anything...

I can see how useful this is, so I'd trade in the 'default layout/layout defined in front matter' functionality for it. I'll take the full power of template inheritance over a nice-looking convenience anytime 👍

Will take care of this today and will add it in the next (0.8.0) release, thanks for bringing it up, for the example, and for your patience 😊
